### PR TITLE
fix(azure AD): fix problem with missing key causing failures in ingestion

### PR DESCRIPTION
**Pull request by anshbansal**
## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

Related issues
- https://github.com/linkedin/datahub/issues/3583

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py [hunk 1]
```diff
@@ -270,16 +270,33 @@ def _get_azure_ad_data(self, kind: str) -> Iterable[List]:
...
-                    "Skipping...".format(azure_ad_group.get("displayName"))
-                )
-                logger.error(error_str)
-                self.report.report_failure("azure_ad_group_mapping", error_str)
+            corp_group_urn, error_str = self._map_identity_to_urn(
+                self._map_azure_ad_group_to_urn,
+                azure_ad_group,
+                "azure_ad_group_mapping",
+            )
+            if error_str is not None:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:299]: Should we log any warning here?

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py [hunk 2]
```diff
@@ -270,16 +270,33 @@ def _get_azure_ad_data(self, kind: str) -> Iterable[List]:
                 self.report.report_failure("_get_azure_ad_data_", error_str)
                 continue
 
+    def _map_identity_to_urn(self, func, id_to_extract, mapping_identifier):
+        result, error_str = None, None
+        try:
+            result = func(id_to_extract)
+        except ValueError as e:
+            error_str = (
+                "Failed to extract DataHub {} from Azure AD User {} due to '{}'".format(
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:279]: Are we always mapping from an Azure AD user? Seems that we can map from groups also

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py [hunk 3]
```diff
@@ -270,16 +270,33 @@ def _get_azure_ad_data(self, kind: str) -> Iterable[List]:
...
+                    mapping_identifier, id_to_extract.get("displayName"), repr(e)
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:280]: Is displayName guaranteed to be there?

**Review by jjoyce0510 — COMMENTED** [review]
In spirit I get this - but it seems a bit overcomplicated / difficult to read. Why not just surround the existing calls in try except?

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py (hunk 1 above)
Comment (anshbansal) [reply metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:299]: The next statement is `logger.error(error_str)` so we are logging error

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py (hunk 2 above)
Comment (anshbansal) [reply metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:279]: this got left out. will fix

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py [hunk 4]
```diff
@@ -270,16 +270,33 @@ def _get_azure_ad_data(self, kind: str) -> Iterable[List]:
                 self.report.report_failure("_get_azure_ad_data_", error_str)
                 continue
 
+    def _map_identity_to_urn(self, func, id_to_extract, mapping_identifier):
+        result, error_str = None, None
+        try:
+            result = func(id_to_extract)
+        except ValueError as e:
```
[the remark this answers is not in the document]
Comment (anshbansal) [reply metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:277]: will do

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py [hunk 5]
```diff
@@ -270,16 +270,32 @@ def _get_azure_ad_data(self, kind: str) -> Iterable[List]:
                 self.report.report_failure("_get_azure_ad_data_", error_str)
                 continue
 
+    def _map_identity_to_urn(self, func, id_to_extract, mapping_identifier, id_type):
+        result, error_str = None, None
+        try:
+            result = func(id_to_extract)
+        except Exception as e:
+            error_str = "Failed to extract DataHub {} from Azure AD {} with name {} due to '{}'".format(
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:278]: `due to '{}'"` is the `'{}'` intentional?

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py [hunk 6]
```diff
@@ -270,16 +270,32 @@ def _get_azure_ad_data(self, kind: str) -> Iterable[List]:
...
+                id_type, id_type, id_to_extract.get("displayName"), repr(e)
+            )
+        if not result:
+            error_str = "Failed to extract DataHub {} from Azure AD {} with name {} due to unkown reason".format(
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:282]: typo `unkown`

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py (hunk 1 above)
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:298]: No - that line got removed in the diff right?

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py (hunk 5 above)
Comment (anshbansal) [reply metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:278]: yes. As that was a separate sentence coming from the exception so I added it.

**Review by jjoyce0510 — APPROVED** [review]
LGTM

**Comment by iasoon** [comment]
I ran into the same issue, this PR did indeed resolve it. Thank you @anshbansal !

**Review by shirshanka — APPROVED** [review]

**Pull request merged**